src/lib/arch/riscv/Instruction_execute.cc

```diff
@@ -827,6 +903,787 @@ void Instruction::execute() {
+
+      // Currently do not read-modify-write ATOMICALLY
+      // Left mostly unimplemented due to Capstone being unable to disassemble
+      // CSR addresses. Some partial functionality is implemented for
+      // correctness of other extensions
+    case Opcode::RISCV_CSRRW: {  // CSRRW rd,csr,rs1
+
+      if (metadata.operands[1].reg == 2) {
+        // Update CPP rounding mode but not floating point CSR as currently no
+        // implementation
+
+        // TODO this won't work properly in OoO core as rounding mode could be
+        // updated before instructions earlier in program order execute
+        // causing them to be rounded incorrectly. Rounding mode needs to be
+        // set on commit and all subsequent instructions must be performed
+        // with updated mode. e.g. this should be performed atomically
+
+        switch (operands[0].get<uint64_t>()) {
+          case 0:
+            fesetround(FE_TONEAREST);
+            break;
+          case 1:
+            fesetround(FE_TOWARDZERO);
+            break;
+          case 2:
+            fesetround(FE_DOWNWARD);
+            break;
+          case 3:
+            fesetround(FE_UPWARD);
+            break;
+          case 4:
+            fesetround(FE_TONEAREST);
+            break;
+          default:
+            // Invalid Case
+            // TODO any subsequent attempt to execute a floating-point
+            // operation with a dynamic rounding mode will raise an illegal
+            // instruction exception.
+            // Should be allowed to be set incorrectly and only caught when
+            // used. Currently SimEng halts on incorrect setting
+            std::cerr << "[SimEng:Instruction_execute] Invalid rounding mode"
+                      << std::endl;
+            exceptionEncountered_ = true;
+            exception_ = InstructionException::IllegalInstruction;
+            break;
+        }
+      }
+
+      results[0] = RegisterValue(0, 8);
+      break;
+    }
+    case Opcode::RISCV_CSRRWI: {  // CSRRWI rd,csr,imm
+      executionNYI();
+      break;
+    }
+    case Opcode::RISCV_CSRRS: {  // CSRRS rd,csr,rs1
+      // dummy implementation
+      results[0] = RegisterValue(static_cast<uint64_t>(0), 8);
+      break;
+    }
+    case Opcode::RISCV_CSRRSI: {  // CSRRSI rd,csr,imm
+      executionNYI();
+      break;
+    }
+    case Opcode::RISCV_CSRRC: {  // CSRRC rd,csr,rs1
+      executionNYI();
+      break;
+    }
+    case Opcode::RISCV_CSRRCI: {  // CSRRCI rd,csr,imm
+      executionNYI();
+      break;
+    }
+
+      // TODO "Apart from transfer operations ... all other floating-point
+      // operations on narrower n-bit operations, n < FLEN, check if the input
+      // operands are correctly NaN-boxed, i.e., all upper FLEN−n bits are 1.
+      // If so, the n least-significant bits of the input are used as the
+      // input value, otherwise the input value is treated as an n-bit
+      // canonical NaN."
```

Comment:
This is another quote from the spec which talks about functionality that is needed for correctness but an implementation could be slow (and possibly never used if the code being run is "correct", which I believe we do assume?). I didn't know exactly how the rest of the F/D extension would look at the time of adding this comment so it hasn't been tackled. I'll give an implementation a go again and we can see if we want to keep it. I don't think it will have too much impact on performance